crates/synthesis/src/qsd.rs

```diff
@@ -0,0 +1,800 @@
+    decompose_uc_rotations(angles, 0, angles.len(), false);
+    for (i, angle) in angles.iter().enumerate() {
+        if angle.abs() > EPS {
+            let _ = out.push_standard_gate(StandardGate::RZ, &[Param::Float(*angle)], &[q_target]);
+        }
+        if i != angles.len() - 1 {
+            // Is this a correct translation from python?
+            // binary_rep = np.binary_repr(i + 1)
+            // q_contr_index = len(binary_rep) - len(binary_rep.rstrip("0"))
+            let q_ctrl_index = (i + 1).trailing_zeros();
+            let _ = out.push_standard_gate(
+                StandardGate::CX,
+                &[],
+                &[q_controls[q_ctrl_index as usize], q_target],
+            );
+        } else if vw_type_all {
+            let q_ctrl_index = num_qubits - 2;
+            let _ = out.push_standard_gate(
+                StandardGate::CX,
+                &[],
+                &[q_controls[q_ctrl_index], q_target],
+            );
+        };
+    }
+    Ok(out)
+}
+
+/// Calculates rotation angles for a uniformly controlled R_t gate with a C-NOT gate at
+/// the end of the circuit. The rotation angles of the gate R_t are stored in
+/// angles[start_index:end_index]. If reversed_dec == True, it decomposes the gate such that
+/// there is a C-NOT gate at the start of the circuit (in fact, the circuit topology for
+/// the reversed decomposition is the reversed one of the original decomposition)
+fn decompose_uc_rotations(
+    angles: &mut [f64],
+    start_index: usize,
+    end_index: usize,
+    reversed_decomposition: bool,
+) {
+    let interval_len_half = (end_index - start_index) / 2;
+    for i in start_index..start_index + interval_len_half {
+        if !reversed_decomposition {
+            let new_angles = update_angle(angles[i], angles[i + interval_len_half]);
+            angles[i] = new_angles[0];
+            angles[i + interval_len_half] = new_angles[1];
+        } else {
+            let new_angles = update_angle(angles[i], angles[i + interval_len_half]);
+            angles[i + interval_len_half] = new_angles[0];
+            angles[i] = new_angles[1];
+        }
+    }
+    if interval_len_half > 1 {
+        decompose_uc_rotations(angles, start_index, start_index + interval_len_half, false);
+        decompose_uc_rotations(angles, start_index + interval_len_half, end_index, true);
+    }
+}
+
+fn update_angle(angle_1: f64, angle_2: f64) -> [f64; 2] {
+    [(angle_1 + angle_2) / 2., (angle_1 - angle_2) / 2.]
+}
+
+fn append(circ: &mut CircuitData, new: CircuitData, qubit_map: &[Qubit]) -> PyResult<()> {
+    let mut new_qubits_map: HashMap<Interned<[Qubit]>, Vec<Qubit>> = HashMap::new();
+    for inst in new.data() {
+        let qubits_map = new_qubits_map.entry(inst.qubits).or_insert_with(|| {
+            new.get_qargs(inst.qubits)
+                .iter()
+                .map(|x| qubit_map[x.index()])
+                .collect()
+        });
```

Comment:
I feel like there is a better way to do this that doesn't involve so many allocations. But I need to think about how to reuse and map the interner indices more directly and check the api for the interners to figure out the correct method to do this so I don't have a concrete suggestion, just to point out this is inefficient.

A small incremental improvement here is to at least do
```suggestion
    let mut new_qubits_map: HashMap<Interned<[Qubit]>, Vec<Qubit>> = HashMap::with_capacity(new.qargs_interner().len());
    for inst in new.data() {
        let qubits_map = new_qubits_map.entry(inst.qubits).or_insert_with(|| {
            new.get_qargs(inst.qubits)
                .iter()
                .map(|x| qubit_map[x.index()])
                .collect()
        });
```
to avoid needing to grow the hashmap repeatedly.

Reply:
Am I incorrect in thinking we have one allocation per (mapped) qubit vector used in the instructions, which we must create anyway in order to add to the circuit?

Reply:
The instructions stored in the circuit used interned indices and represent the qubit (and clbit) arguments as a single u32 internally, which is typed as `Interned<[Qubit]>`. That's what `get_qargs` takes in. The issue is that the interner  (the data structure that stores the mapping of qarga to interned value) and by extension the interned value are tied to each individual circuit and not portable between circuits. So you need to add  the interner mappings from `new` here with the caller supplied `qubit_map` applied to `self``s interner. In my head there should be a method on the interners that does that bulk add and then you are just mapping old interned value to the new interned value.

Reply:
So I guess we leave this as-is for now?

Reply:
I took care of this in: https://github.com/Qiskit/qiskit/pull/14797/commits/7a80f6c1822c01194fe90847afde328acef6a900